javaagent-tooling/src/main/java/io/opentelemetry/javaagent/tooling/instrumentation/InstrumentationModuleInstaller.java

```diff
@@ -69,44 +72,87 @@ AgentBuilder install(
     }
 
     if (instrumentationModule.isIndyModule()) {
-      return installIndyModule(instrumentationModule, parentAgentBuilder);
+      return installIndyModule(instrumentationModule, parentAgentBuilder, config);
     } else {
       return installInjectingModule(instrumentationModule, parentAgentBuilder, config);
     }
   }
 
   private AgentBuilder installIndyModule(
-      InstrumentationModule instrumentationModule, AgentBuilder parentAgentBuilder) {
-
-    IndyModuleRegistry.registerIndyModule(instrumentationModule);
-
+      InstrumentationModule instrumentationModule,
+      AgentBuilder parentAgentBuilder,
+      ConfigProperties config) {
+    List<String> helperClassNames =
+        InstrumentationModuleMuzzle.getHelperClassNames(instrumentationModule);
     HelperResourceBuilderImpl helperResourceBuilder = new HelperResourceBuilderImpl();
     instrumentationModule.registerHelperResources(helperResourceBuilder);
+    List<TypeInstrumentation> typeInstrumentations = instrumentationModule.typeInstrumentations();
+    if (typeInstrumentations.isEmpty()) {
+      if (!helperClassNames.isEmpty() || !helperResourceBuilder.getResources().isEmpty()) {
+        logger.log(
+            WARNING,
+            "Helper classes and resources won't be injected if no types are instrumented: {0}",
+            instrumentationModule.instrumentationName());
+      }
+
+      return parentAgentBuilder;
+    }
+
+    List<String> injectedHelperClassNames = Collections.emptyList();
+    if (instrumentationModule instanceof ExperimentalInstrumentationModule) {
+      ExperimentalInstrumentationModule experimentalInstrumentationModule =
+          (ExperimentalInstrumentationModule) instrumentationModule;
+      injectedHelperClassNames = experimentalInstrumentationModule.injectedClassNames();
+      helperClassNames.removeAll(injectedHelperClassNames);
+    }
+
+    IndyModuleRegistry.registerIndyModule(instrumentationModule);
 
     ClassInjectorImpl injectedClassesCollector = new ClassInjectorImpl(instrumentationModule);
     if (instrumentationModule instanceof ExperimentalInstrumentationModule) {
       ((ExperimentalInstrumentationModule) instrumentationModule)
           .injectClasses(injectedClassesCollector);
     }
 
+    // TODO (Jonas): Adapt MuzzleMatcher to use the same type lookup strategy as the
+    // InstrumentationModuleClassLoader (see IndyModuleTypePool)
+    MuzzleMatcher muzzleMatcher =
+        new MuzzleMatcher(logger, instrumentationModule, config) {
+          @Override
+          public boolean matches(
+              TypeDescription typeDescription,
+              ClassLoader classLoader,
+              JavaModule module,
+              Class<?> classBeingRedefined,
+              ProtectionDomain protectionDomain) {
+            ClassLoader instrumentationClassLoader =
```

Comment:
If my understanding of `AgentCachingPoolStrategy` is correct, our TypeDescription-Cache ensures that we never hold a strong reference to classloaders, correct?

As a result, the caching won't really work here and it will also cause good entries to be evicted from the cache if it is fixed-size due to "flooding" the cache with temporary classloader.

I think we could avoid these problems and keep the muzzle-matching efficient by instead altering `ReferenceMatcher.createTypePool`:

```java
// loader cannot be null, must pass "bootstrap proxy" instead of bootstrap class loader
  private TypePool createTypePool(ClassLoader loader) {
    // ok to use locationStrategy() without fallback bootstrap proxy here since loader is non-null
    TypePool instrumentedClPool = AgentTooling.poolStrategy()
        .typePool(AgentTooling.locationStrategy().classFileLocator(loader), loader);
    if(instrumentationModule.isIndy()) {
      TypePool instrumentedClPool = AgentTooling.poolStrategy()
          .typePool(AgentTooling.locationStrategy().classFileLocator(loader), loader);
      ClassLoader moduleOrAgentCl = instrumentationModule.getClass().getClassLoader();
      TypePool agentOrExtensionPool = AgentTooling.poolStrategy()
          .typePool(AgentTooling.locationStrategy().classFileLocator(moduleOrAgentCl), moduleOrAgentCl);
      return new TypePool() {
        @Override
        public Resolution describe(String name) {
          if(name.startsWith("io.opentelemetry.javaagent")) {
            return agentOrExtensionPool.describe(name);
          }
          return instrumentedClPool.describe(name);
        }

        @Override
        public void clear() {
          instrumentedClPool.clear();
          agentOrExtensionPool.clear();
        }
      };
    }
    return instrumentedClPool;
  }
```

Reply:
This will require further investigation. As far as I understand we don't pass the temporary class loader here but the real instrumentation loader. Keep in mind that having to transform a class is really a special case, relatively few classes get transformed. It should be fine as long as loading classes that don't get transformed does minimal work.

Reply:
> This will require further investigation. As far as I understand we don't pass the temporary class loader here but the real instrumentation loader.

Correct, but we make sure that we never hold a strong reference to that `InstrumentationModuleClassloader`, because that would result in a strong reference to the instrumented application classloader.

The `InstrumentationModuleClassloader` will only be strongly-referenced when the first `invokedynamic`-instruction is linked, therefore it may be GCed until this happens.

> Keep in mind that having to transform a class is really a special case, relatively few classes get transformed. It should be fine as long as loading classes that don't get transformed does minimal work.

Makes sense, so we can keep it as is for now. I just thought that my proposed alternative with the custom `TypePool` might avoid these issues altogether, but we don't need to do this now if it is too much effort.

Reply:
Would it make sense to insert a strong reference from application class loader to `InstrumentationModuleClassloader` so that `InstrumentationModuleClassloader` couldn't be gcd before the application class loader? Could do this be defining a class in application class loader that could keep the reference in a static field or map.

Reply:
I don't think that this is worth the effort and complexity now, because IINM the custom TypePool I suggested would still be better and imo less complex: When caching `TypeDescriptions` the corresponding classloader is used as part of the cache-key. If we use the `InstrumentationModuleClassloader` here, this will only allow the cache to be effective for classes from the same classloader when checking for the same `InstrumentationModule`.

If we instead use something like the custom type pool I suggested, the `TypeDescriptions` will instead be associated with either the application classloader or the agent/extension classloader, depending where they come from, allowing a better "reuse" in theory.

I think for now the "risk" of the InstrumentationModuleCL being GCed here is okay. If you want to, I can do a follow-up PR trying to implement that custom type pool.